lib/vrl/stdlib/src/tag_types_externally.rs

```diff
@@ -0,0 +1,253 @@
+use shared::btreemap;
+use std::collections::BTreeMap;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct TagTypesExternally;
+
+impl Function for TagTypesExternally {
+    fn identifier(&self) -> &'static str {
+        "tag_types_externally"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "object",
+            source: indoc! {r#"
+                tag_types_externally({
+                    "message": "Hello world",
+                    "request": {
+                        "duration_ms": 67.9
+                    }
+                })
+            "#},
+            result: Ok(
+                r#"{ "message": { "bytes": "Hello world" }, "request": { "duration_ms": { "float": 67.9 } } }"#,
+            ),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(TagTypesExternallyFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ANY,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TagTypesExternallyFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for TagTypesExternallyFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let tagged_externally = tag_type_externally(&value);
+
+        Ok(tagged_externally)
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new()
+            .infallible()
+            .object::<(), Kind>(map! {
+                (): Kind::all()
+            })
+            .add_array_mapped::<(), Kind>(map! {
+                (): Kind::all()
+            })
```

Comment:
I see, then it would be good to (1) add an `Example` demonstrating this, and (2) update the docs to mention it can also return an array.

Reply:
Also, in that case, shouldn't we add a conditional here, to check the type of `value` and define the resulting type def based on its type def?

That would allow for more precise type definitions when calling this function.